pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java

```diff
@@ -220,13 +232,25 @@ public String getPaddedString(int index, int numBytesPerValue, byte[] buffer) {
     // To get the length of the byte array, we use the next byte array offset.
     int length = _dataBuffer.getInt(_dataSectionStartOffSet + Integer.BYTES * (index + 1)) - offset;
 
+    // Since the byte[] is returned to the caller and caller could hold a reference to it,
+    // the buffer can't really be reused here. Hence, don't use ThreadLocal buffer in this case.
+    return getBytes(buffer, offset, length, false);
+  }
+
+  private byte[] getBytes(byte[] buffer, int offset, int length, boolean useThreadLocalBuffer) {
     byte[] b;
-    // If the caller didn't pass a buffer, create one with exact length.
-    if (buffer == null) {
-      b = new byte[length];
+    // If the caller has passed a buffer that can be used, use it.
+    if (buffer != null && buffer.length == length) {
+      b = buffer;
+    } else if (useThreadLocalBuffer) {
+      // Check if the current instance of ThreadLocal buffer is big enough. If not, resize it to double the size.
+      while (_reusableBytes.get().length < length) {
+        _reusableBytes.set(new byte[_reusableBytes.get().length * 2]);
```

Comment:
Directly get the final size (next power of 2 for length) instead of doing this while loop

Reply:
Done.